pkg/controllers/edgeconnect/controller.go

```diff
@@ -152,8 +229,244 @@ func (controller *Controller) updateEdgeConnectStatus(ctx context.Context, edgeC
+
+	if err != nil {
+		log.Info("could not create or update deployment for EdgeConnect", "name", desiredDeployment.Name)
+		return err
+	}
+	return nil
+}
+
+func (controller *Controller) reconcileEdgeConnectProvisioner(ctx context.Context, edgeConnect *edgeconnectv1alpha1.EdgeConnect) error { //nolint: revive
+	log.Info("reconcileEdgeConnectProvisioner")
+
+	edgeConnectClient, err := controller.buildEdgeConnectClient(ctx, edgeConnect)
+	if err != nil {
+		return err
+	}
+
+	edgeConnectId, err := getEdgeConnectIdByName(edgeConnectClient, edgeConnect.Name)
+	if err != nil {
+		return err
+	}
+
+	edgeConnectIdFromSecret, err := controller.getEdgeConnectIdFromClientSecret(ctx, edgeConnect)
+	if err != nil {
+		return err
+	}
+
+	if edgeConnectId != "" {
+		if edgeConnectIdFromSecret == "" {
+			log.Info("EdgeConnect has to be recreated due to missing secret")
+			// TODO: managedByDynatraceOperator attribute needed
+			if err := edgeConnectClient.DeleteEdgeConnect(edgeConnectId); err != nil {
+				return err
+			}
+			edgeConnectId = ""
+		} else if edgeConnectId != edgeConnectIdFromSecret {
+			// TODO: managedByDynatraceOperator attribute needed
+			log.Info("EdgeConnect has to be recreated due to invalid Id")
+			if err := edgeConnectClient.DeleteEdgeConnect(edgeConnectId); err != nil {
+				return err
+			}
+			edgeConnectId = ""
+		}
+	}
+
+	if edgeConnectId == "" {
+		err := controller.createEdgeConnect(ctx, edgeConnectClient, edgeConnect)
+		if err != nil {
+			return err
+		}
+		return controller.createOrUpdateEdgeConnectDeployment(ctx, edgeConnect)
+	}
+
+	err = controller.updateEdgeConnect(ctx, edgeConnectClient, edgeConnect)
+	if err != nil {
+		return err
+	}
+	return controller.createOrUpdateEdgeConnectDeployment(ctx, edgeConnect)
+}
+
+func (controller *Controller) buildEdgeConnectClient(ctx context.Context, edgeConnect *edgeconnectv1alpha1.EdgeConnect) (edgeconnect.Client, error) {
+	oauthClientId, oauthClientSecret, err := controller.getOauthCredentials(ctx, edgeConnect)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	return controller.edgeConnectClientBuilder(ctx, edgeConnect, oauthClientId, oauthClientSecret)
+}
+
+func (controller *Controller) getOauthCredentials(ctx context.Context, edgeConnect *edgeconnectv1alpha1.EdgeConnect) (string, string, error) {
+	query := kubeobjects.NewSecretQuery(ctx, controller.client, controller.apiReader, log)
+	secret, err := query.Get(types.NamespacedName{
+		Name:      edgeConnect.Spec.OAuth.ClientSecret,
+		Namespace: edgeConnect.Namespace,
+	})
+	if err != nil {
+		return "", "", errors.WithStack(err)
+	}
+
+	oauthClientId, err := kubeobjects.ExtractToken(&secret, consts.KeyEdgeConnectOauthClientID)
+	if err != nil {
+		return "", "", errors.WithStack(err)
+	}
+
+	oauthClientSecret, err := kubeobjects.ExtractToken(&secret, consts.KeyEdgeConnectOauthClientSecret)
+	if err != nil {
+		return "", "", errors.WithStack(err)
+	}
+	return oauthClientId, oauthClientSecret, nil
```

Comment:
Since we need these two as pairs, we can create a new type to hold both; it will simplify the return type and reduce the amount of func arguments later from 2 to 1.